File: integration-tests/csharp/Tests/UnionsTests.cs
// Copyright 2026 KONGSBERG
//
// Redistribution and use in source and binary forms, with or without
// modification, are permitted provided that the following conditions are met:
//
// 1. Redistributions of source code must retain the above copyright notice,
//    this list of conditions and the following disclaimer.
//
// 2. Redistributions in binary form must reproduce the above copyright notice,
//    this list of conditions and the following disclaimer in the documentation
//    and/or other materials provided with the distribution.
//
// 3. Neither the name of the copyright holder nor the names of its contributors
//    may be used to endorse or promote products derived from this software
//    without specific prior written permission.
//
// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND
// ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
// WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
// DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE
// FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
// DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
// SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
// CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
// OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
// OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.

using Xunit;
using UnionTypes;

namespace IntegrationTests;

public class UnionsTests
{
    [Fact]
    public void IntOrString_DefaultState()
    {
        var u = new IntOrString();
        Assert.Equal(0, u.Discriminator);
    }

    [Fact]
    public void IntOrString_SetIntVal()
    {
        var u = new IntOrString();
        u.IntVal = 42;
        Assert.Equal(1, u.Discriminator);
        Assert.Equal(42, u.IntVal);
    }

    [Fact]
    public void IntOrString_SetStrVal()
    {
        var u = new IntOrString();
        u.StrVal = "hello";
        Assert.Equal(2, u.Discriminator);
        Assert.Equal("hello", u.StrVal);
    }

    [Fact]
    public void IntOrString_AccessWrongMemberThrows()
    {
        var u = new IntOrString();
        u.IntVal = 42;
        Assert.Throws<InvalidOperationException>(() => u.StrVal);
    }

    [Fact]
    public void IntOrString_SwitchingVariantClearsOther()
    {
        var u = new IntOrString();
        u.IntVal = 42;
        Assert.Equal(1, u.Discriminator);

        u.StrVal = "hello";
        Assert.Equal(2, u.Discriminator);
        Assert.Equal("hello", u.StrVal);
        Assert.Throws<InvalidOperationException>(() => u.IntVal);
    }

    [Fact]
    public void IntOrString_CopyConstructor()
    {
        var u1 = new IntOrString();
        u1.IntVal = 99;
        var u2 = new IntOrString(u1);
        Assert.Equal(u1.Discriminator, u2.Discriminator);
        Assert.Equal(u1.IntVal, u2.IntVal);
    }

    [Fact]
    public void IntOrString_Equality()
    {
        var u1 = new IntOrString();
        var u2 = new IntOrString();
        u1.IntVal = 42;
        u2.IntVal = 42;
        Assert.Equal(u1, u2);

        u2.IntVal = 100;
        Assert.NotEqual(u1, u2);
    }

    [Fact]
    public void TypedValue_EnumDiscriminator()
    {
        var v = new TypedValue();
        v.IntValue = 123;
        Assert.Equal(ValueKind.IntKind, v.Discriminator);
        Assert.Equal(123, v.IntValue);
    }

    [Fact]
    public void TypedValue_FloatValue()
    {
        var v = new TypedValue();
        v.FloatValue = 3.14;
        Assert.Equal(ValueKind.FloatKind, v.Discriminator);
        Assert.Equal(3.14, v.FloatValue);
    }

    [Fact]
    public void TypedValue_StringValue()
    {
        var v = new TypedValue();
        v.StringValue = "test";
        Assert.Equal(ValueKind.StringKind, v.Discriminator);
        Assert.Equal("test", v.StringValue);
    }

    [Fact]
    public void BoolSwitch_TrueCase()
    {
        var b = new BoolSwitch();
        b.TrueVal = 42;
        Assert.True(b.Discriminator);
        Assert.Equal(42, b.TrueVal);
    }

    [Fact]
    public void BoolSwitch_FalseCase()
    {
        var b = new BoolSwitch();
        b.FalseVal = "hello";
        Assert.False(b.Discriminator);
        Assert.Equal("hello", b.FalseVal);
    }

    [Fact]
    public void MultiCase_SmallValWithSetMethod()
    {
        var m = new MultiCase();
        m.SetSmallVal(10, 2);
        Assert.Equal(2, m.Discriminator);
        Assert.Equal(10, m.SmallVal);
    }

    [Fact]
    public void MultiCase_SmallValMultipleDiscriminators()
    {
        var m = new MultiCase();
        m.SetSmallVal(5, 1);
        Assert.Equal(1, m.Discriminator);
        Assert.Equal(5, m.SmallVal);

        m.SetSmallVal(6, 3);
        Assert.Equal(3, m.Discriminator);
        Assert.Equal(6, m.SmallVal);
    }

    [Fact]
    public void MultiCase_InvalidDiscriminatorThrows()
    {
        var m = new MultiCase();
        Assert.Throws<ArgumentException>(() => m.SetSmallVal(10, 5));
    }

    [Fact]
    public void MultiCase_TextVal()
    {
        var m = new MultiCase();
        m.SetTextVal("test", 10);
        Assert.Equal(10, m.Discriminator);
        Assert.Equal("test", m.TextVal);

        m.SetTextVal("test2", 20);
        Assert.Equal(20, m.Discriminator);
        Assert.Equal("test2", m.TextVal);
    }

    [Fact]
    public void IntOrString_DefaultVal()
    {
        var u = new IntOrString();
        u.DefaultVal = true;
        Assert.Equal(0, u.Discriminator);
        Assert.True(u.DefaultVal);
        Assert.Throws<InvalidOperationException>(() => u.IntVal);
        Assert.Throws<InvalidOperationException>(() => u.StrVal);
    }
}

File: integration-tests/java/src/test/java/integrationtests/UnionsTests.java
// Copyright 2026 KONGSBERG
//
// Redistribution and use in source and binary forms, with or without
// modification, are permitted provided that the following conditions are met:
//
// 1. Redistributions of source code must retain the above copyright notice,
//    this list of conditions and the following disclaimer.
//
// 2. Redistributions in binary form must reproduce the above copyright notice,
//    this list of conditions and the following disclaimer in the documentation
//    and/or other materials provided with the distribution.
//
// 3. Neither the name of the copyright holder nor the names of its contributors
//    may be used to endorse or promote products derived from this software
//    without specific prior written permission.
//
// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND
// ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
// WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
// DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE
// FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
// DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
// SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
// CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
// OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
// OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.

package integrationtests;

import static org.junit.jupiter.api.Assertions.*;

import org.junit.jupiter.api.Test;
import union_types.*;

class UnionsTests {

    @Test
    void intOrString_defaultsToDefaultVal() {
        var u = new IntOrString();
        assertEquals(0, u.getDiscriminator());
        assertFalse(u.getDefaultVal());
    }

    @Test
    void defaultDiscriminatorCase_defaultsToValue() {
        var u = new DefaultDiscriminatorCase();
        assertEquals(0, u.getDiscriminator());
        assertEquals(0, u.getValue());
    }

    @Test
    void intOrString_setIntVal() {
        var u = new IntOrString();
        u.setIntVal(42);
        assertEquals(1, u.getDiscriminator());
        assertEquals(42, u.getIntVal());
    }

    @Test
    void intOrString_setStrVal() {
        var u = new IntOrString();
        u.setStrVal("hello");
        assertEquals(2, u.getDiscriminator());
        assertEquals("hello", u.getStrVal());
    }

    @Test
    void intOrString_setDefaultVal() {
        var u = new IntOrString();
        u.setIntVal(42);
        u.setDefaultVal(true);
        assertEquals(0, u.getDiscriminator());
        assertTrue(u.getDefaultVal());
    }

    @Test
    void intOrString_wrongAccessThrows() {
        var u = new IntOrString();
        u.setStrVal("test");
        assertThrows(IllegalStateException.class, () -> u.getIntVal());
    }

    @Test
    void intOrString_copyConstructor() {
        var u1 = new IntOrString();
        u1.setStrVal("copy me");
        var u2 = new IntOrString(u1);
        assertEquals(u1.getDiscriminator(), u2.getDiscriminator());
        assertEquals(u1.getStrVal(), u2.getStrVal());
        assertNotSame(u1, u2);
    }

    @Test
    void intOrString_equality() {
        var u1 = new IntOrString();
        u1.setIntVal(100);
        var u2 = new IntOrString();
        u2.setIntVal(100);
        var u3 = new IntOrString();
        u3.setIntVal(200);
        assertEquals(u1, u2);
        assertNotEquals(u1, u3);
    }

    @Test
    void intOrString_hashCodeConsistent() {
        var u1 = new IntOrString();
        u1.setIntVal(42);
        var u2 = new IntOrString();
        u2.setIntVal(42);
        assertEquals(u1.hashCode(), u2.hashCode());
    }

    @Test
    void intOrString_clone() {
        var u1 = new IntOrString();
        u1.setStrVal("original");
        var u2 = u1.clone();
        assertEquals(u1, u2);
        assertNotSame(u1, u2);
    }

    @Test
    void boolSwitch_defaultsToFalse() {
        var u = new BoolSwitch();
        assertFalse(u.getDiscriminator());
        assertEquals("", u.getFalseVal());
    }

    @Test
    void boolSwitch_setTrueVal() {
        var u = new BoolSwitch();
        u.setTrueVal(123);
        assertTrue(u.getDiscriminator());
        assertEquals(123, u.getTrueVal());
    }

    @Test
    void boolSwitch_setFalseVal() {
        var u = new BoolSwitch();
        u.setFalseVal("false branch");
        assertFalse(u.getDiscriminator());
        assertEquals("false branch", u.getFalseVal());
    }

    @Test
    void boolSwitch_wrongAccessThrows() {
        var u = new BoolSwitch();
        u.setTrueVal(1);
        assertThrows(IllegalStateException.class, () -> u.getFalseVal());
    }

    @Test
    void boolSwitch_equality() {
        var u1 = new BoolSwitch();
        u1.setFalseVal("test");
        var u2 = new BoolSwitch();
        u2.setFalseVal("test");
        assertEquals(u1, u2);
    }

    @Test
    void typedValue_defaultsToIntKind() {
        var u = new TypedValue();
        assertEquals(ValueKind.INT_KIND, u.getDiscriminator());
        assertEquals(0, u.getIntValue());
    }

    @Test
    void typedValue_setIntValue() {
        var u = new TypedValue();
        u.setIntValue(42);
        assertEquals(ValueKind.INT_KIND, u.getDiscriminator());
        assertEquals(42, u.getIntValue());
    }

    @Test
    void typedValue_setFloatValue() {
        var u = new TypedValue();
        u.setFloatValue(3.14);
        assertEquals(ValueKind.FLOAT_KIND, u.getDiscriminator());
        assertEquals(3.14, u.getFloatValue(), 0.001);
    }

    @Test
    void typedValue_setStringValue() {
        var u = new TypedValue();
        u.setStringValue("hello");
        assertEquals(ValueKind.STRING_KIND, u.getDiscriminator());
        assertEquals("hello", u.getStringValue());
    }

    @Test
    void typedValue_wrongAccessThrows() {
        var u = new TypedValue();
        u.setIntValue(1);
        assertThrows(IllegalStateException.class, () -> u.getFloatValue());
        assertThrows(IllegalStateException.class, () -> u.getStringValue());
    }

    @Test
    void typedValue_equality() {
        var u1 = new TypedValue();
        u1.setFloatValue(2.718);
        var u2 = new TypedValue();
        u2.setFloatValue(2.718);
        assertEquals(u1, u2);
    }

    @Test
    void multiCase_defaultsToFlag() {
        var u = new MultiCase();
        assertEquals(0, u.getDiscriminator());
        assertFalse(u.getFlag());
    }

    @Test
    void multiCase_setSmallVal() {
        var u = new MultiCase();
        u.setSmallVal(42);
        assertEquals(1, u.getDiscriminator());
        assertEquals(42, u.getSmallVal());
    }

    @Test
    void multiCase_setSmallValWithDiscriminator() {
        var u = new MultiCase();
        u.setSmallVal(99, 2);
        assertEquals(2, u.getDiscriminator());
        assertEquals(99, u.getSmallVal());
    }

    @Test
    void multiCase_setSmallValWithDiscriminator3() {
        var u = new MultiCase();
        u.setSmallVal(77, 3);
        assertEquals(3, u.getDiscriminator());
        assertEquals(77, u.getSmallVal());
    }

    @Test
    void multiCase_setTextVal() {
        var u = new MultiCase();
        u.setTextVal("hello");
        assertEquals(10, u.getDiscriminator());
        assertEquals("hello", u.getTextVal());
    }

    @Test
    void multiCase_setTextValWithDiscriminator() {
        var u = new MultiCase();
        u.setTextVal("world", 20);
        assertEquals(20, u.getDiscriminator());
        assertEquals("world", u.getTextVal());
    }

    @Test
    void multiCase_setFlag() {
        var u = new MultiCase();
        u.setFlag(true);
        assertEquals(0, u.getDiscriminator());
        assertTrue(u.getFlag());
    }

    @Test
    void multiCase_wrongAccessThrows() {
        var u = new MultiCase();
        u.setSmallVal(1);
        assertThrows(IllegalStateException.class, () -> u.getTextVal());
        assertThrows(IllegalStateException.class, () -> u.getFlag());
    }

    @Test
    void multiCase_invalidDiscriminatorForSmallValThrows() {
        var u = new MultiCase();
        assertThrows(IllegalStateException.class, () -> u.setSmallVal(1, 5));
    }

    @Test
    void multiCase_equality() {
        var u1 = new MultiCase();
        u1.setTextVal("test", 20);
        var u2 = new MultiCase();
        u2.setTextVal("test", 20);
        var u3 = new MultiCase();
        u3.setTextVal("test", 10);
        assertEquals(u1, u2);
        assertNotEquals(u1, u3);
    }
}
